File: src/features/auth/auth.service.ts
import { Injectable, UnauthorizedException, ForbiddenException, BadRequestException } from '@nestjs/common';
import * as argon2 from 'argon2';
import * as crypto from 'crypto';
import { AuditLogService } from '../../common/services/audit-log.service';
import { ActorContext, UsersService } from '../users/users.service';
import { OrganizationsService } from '../organizations/organizations.service';
import { RegisterDto } from './dto/register.dto';
import { VerifyEmailDto } from './dto/verify-email.dto';
import { ForgotPasswordDto } from './dto/forgot-password.dto';
import { ResetPasswordDto } from './dto/reset-password.dto';
import { EmailService } from '../email/email.service';
import { Role } from '../../common/roles';
import { STRONG_PASSWORD_REGEX } from './dto/password-rules';
import zxcvbn from 'zxcvbn';
import { WaitlistService } from '../waitlist/waitlist.service';
import { normalizeDomainFromEmail } from '../../common/utils/domain.util';
@Injectable()
export class AuthService {
  constructor(
    private readonly audit: AuditLogService,
    private readonly users: UsersService,
    private readonly orgs: OrganizationsService,
    private readonly email: EmailService,
    private readonly waitlist: WaitlistService
  ) {}

  private generateToken(ttlMs: number) {
    const token = crypto.randomBytes(32).toString('hex');
    const hash = crypto.createHash('sha256').update(token).digest('hex');
    const expires = new Date(Date.now() + ttlMs);
    return { token, hash, expires };
  }

  async login(email: string, pass: string) {
    const user = await this.users.findByEmail(email);
    const isValid = user && (await argon2.verify(user.passwordHash, pass));
    await this.audit.log({
      eventType: 'auth.login',
      userId: user?.id,
      orgId: user?.organizationId,
      metadata: { email, success: !!isValid },
    });
    if (!isValid) throw new UnauthorizedException();
    if (user.archivedAt) throw new ForbiddenException('User archived');
    if (user.legalHold) throw new ForbiddenException('User on legal hold');
    if (!user.isEmailVerified) throw new ForbiddenException('Email not verified');

    const updated = await this.users.markLastLogin(user.id);
    return this.sanitizeUser(updated || user);
  }

  private sanitizeUser(user: any) {
    const { passwordHash, ...rest } = user.toObject ? user.toObject() : user;
    return rest;
  }

  passwordStrength(password: string) {
    const result = zxcvbn(password || '');
    return {
      score: result.score,
      crackTimesDisplay: result.crack_times_display,
      feedback: result.feedback,
      guessesLog10: result.guesses_log10,
    };
  }

  async listUsers(orgId?: string, actor?: ActorContext) {
    const resolvedActor = actor || { orgId };
    return this.users.list(resolvedActor, { orgId });
  }

  async register(dto: RegisterDto) {
    if (!STRONG_PASSWORD_REGEX.test(dto.password)) {
      throw new BadRequestException('Password does not meet strength requirements');
    }
    if (!dto.legalAccepted) {
      throw new BadRequestException('You must accept the Privacy Policy and Terms & Conditions to create an account');
    }
    const normalizedEmail = dto.email.toLowerCase();
    const domain = normalizeDomainFromEmail(normalizedEmail);
    if (!domain) throw new BadRequestException('Invalid email domain');

    const firstName = (dto.firstName || '').trim() || undefined;
    const lastName = (dto.lastName || '').trim() || undefined;
    const derivedUsername =
      (dto.username || '').trim() ||
      [firstName, lastName].filter(Boolean).join(' ') ||
      normalizedEmail.split('@')[0];

    const waitlistEntry: any = await this.waitlist.findByEmail(normalizedEmail);
    const inviteGateEnabled = this.waitlist.shouldEnforceInviteGate();
    if (inviteGateEnabled) {
      if (!waitlistEntry || !['invited', 'activated'].includes(waitlistEntry?.status || '')) {
        await this.waitlist.logEvent('waitlist.register_blocked', {
          email: normalizedEmail,
          status: waitlistEntry?.status || 'missing',
        });
        throw new ForbiddenException(
          'Invite required. Request early access, verify your email + phone, and we will invite your company in a wave.'
        );
      }

      const fullyVerified = waitlistEntry?.verifyStatus === 'verified' && waitlistEntry?.phoneVerifyStatus === 'verified';
      if (!fullyVerified) {
        await this.waitlist.logEvent('waitlist.register_blocked_unverified', {
          email: normalizedEmail,
          verifyStatus: waitlistEntry?.verifyStatus,
          phoneVerifyStatus: waitlistEntry?.phoneVerifyStatus,
        });
        throw new ForbiddenException('Please verify your email and phone before creating your account.');
      }

      const inviteToken = (dto as any).inviteToken;
      if (!inviteToken) {
        throw new ForbiddenException('Invite link required. Please use the invite email to register.');
      }
      const inviteHash = crypto.createHash('sha256').update(inviteToken).digest('hex');
      const tokenExpiresAt = waitlistEntry?.inviteTokenExpiresAt ? new Date(waitlistEntry.inviteTokenExpiresAt) : null;
      const tokenExpired = tokenExpiresAt ? tokenExpiresAt.getTime() < Date.now() : false;
      if (!waitlistEntry?.inviteTokenHash || waitlistEntry.inviteTokenHash !== inviteHash || tokenExpired) {
        throw new ForbiddenException('Invite link invalid or expired. Please request a fresh invite.');
      }
    }

    let domainLocked = false;
    if (this.waitlist.domainGateEnabled()) {
      const existingDomainOrg = await this.orgs.findByDomain(domain);
      if (existingDomainOrg) {
        await this.waitlist.logEvent('waitlist.register_blocked_domain', {
          email: normalizedEmail,
          domain,
          orgId: (existingDomainOrg as any)._id || (existingDomainOrg as any).id,
        });
        throw new ForbiddenException('Your company already has access. Please ask your org admin to invite you.');
      }
      const claimed = await this.waitlist.acquireDomainClaim(domain);
      domainLocked = claimed;
      if (!claimed) {
        await this.waitlist.logEvent('waitlist.register_blocked_domain_race', { email: normalizedEmail, domain });
        throw new ForbiddenException('Your company already has access. Please ask your org admin to invite you.');
      }
    }

    try {
      let organizationId = dto.organizationId;
      let createdOrgId: string | undefined;
      if (!organizationId) {
        const orgName = dto.organizationName || `${derivedUsername}'s Organization`;
        const org = await this.orgs.create({ name: orgName, primaryDomain: domain });
        organizationId = org.id || (org as any)._id;
        createdOrgId = organizationId;
      } else {
        await this.orgs.findById(organizationId);
      }

      let user;
      const emailVerification = inviteGateEnabled ? null : this.generateToken(1000 * 60 * 60 * 24); // 24h
      const assignedRole = createdOrgId ? Role.OrgOwner : Role.User;
      user = await this.users.create({
        username: derivedUsername,
        firstName,
        lastName,
        email: normalizedEmail,
        password: dto.password,
        organizationId,
        role: assignedRole,
        isOrgOwner: assignedRole === Role.OrgOwner,
        verificationTokenHash: emailVerification?.hash ?? null,
        verificationTokenExpires: emailVerification?.expires ?? null,
        isEmailVerified: inviteGateEnabled,
      }, undefined, { enforceSeat: true });

      const userId = (user as any).id || (user as any)._id;

      if (createdOrgId) {
        await this.orgs.setOwner(createdOrgId, userId);
      }
      await this.waitlist.markActivated(normalizedEmail, waitlistEntry?.cohortTag);
      await this.audit.log({
        eventType: 'auth.register',
        userId,
        orgId: organizationId,
      });
      if (emailVerification) {
        await this.email.sendVerificationEmail(normalizedEmail, emailVerification.token, organizationId, dto.username);
      }
      return user;
    } catch (err) {
      throw err;
    } finally {
      // ensure lock doesn't linger if creation failed silently
      if (domainLocked) {
        await this.waitlist.releaseDomainClaim(domain);
      }
    }
  }

  async verifyEmail(dto: VerifyEmailDto) {
    const hash = crypto.createHash('sha256').update(dto.token).digest('hex');
    const user = await this.users.findByVerificationToken(hash);
    if (!user) throw new BadRequestException('Invalid or expired token');
    await this.users.clearVerificationToken(user.id);
    await this.audit.log({
      eventType: 'auth.verify_email',
      userId: user.id,
      orgId: user.organizationId,
    });
    return this.sanitizeUser(user);
  }

  async forgotPassword(dto: ForgotPasswordDto) {
    const user = await this.users.findByEmail(dto.email);
    if (!user) return { status: 'ok' }; // do not leak existence
    if (user.legalHold || user.archivedAt) throw new ForbiddenException('User unavailable');
    const { token, hash, expires } = this.generateToken(1000 * 60 * 60); // 1h
    await this.users.setResetToken(user.id, hash, expires);
    await this.audit.log({
      eventType: 'auth.password_reset_requested',
      userId: user.id,
      orgId: user.organizationId,
    });
    await this.email.sendPasswordResetEmail(user.email, token, user.organizationId as string, user.username);
    return { status: 'ok' };
  }

  async resetPassword(dto: ResetPasswordDto) {
    if (!STRONG_PASSWORD_REGEX.test(dto.newPassword)) {
      throw new BadRequestException('Password does not meet strength requirements');
    }
    const hash = crypto.createHash('sha256').update(dto.token).digest('hex');
    const user = await this.users.findByResetToken(hash);
    if (!user) throw new BadRequestException('Invalid or expired token');
    if (user.legalHold || user.archivedAt) throw new ForbiddenException('User unavailable');
    const updated = await this.users.clearResetTokenAndSetPassword(user.id, dto.newPassword);
    await this.audit.log({
      eventType: 'auth.password_reset',
      userId: user.id,
      orgId: user.organizationId,
    });
    return this.sanitizeUser(updated);
  }
}

File: spring-boot-api/src/main/java/com/mytegroup/api/service/auth/AuthService.java
package com.mytegroup.api.service.auth;

import com.mytegroup.api.common.enums.Role;
import com.mytegroup.api.common.util.DomainUtil;
import com.mytegroup.api.common.util.PasswordValidator;
import com.mytegroup.api.common.util.TokenHashUtil;
import com.mytegroup.api.entity.core.Organization;
import com.mytegroup.api.entity.core.User;
import com.mytegroup.api.exception.BadRequestException;
import com.mytegroup.api.exception.ForbiddenException;
import com.mytegroup.api.exception.UnauthorizedException;
import com.mytegroup.api.service.common.AuditLogService;
import com.mytegroup.api.service.email.EmailService;
import com.mytegroup.api.service.organizations.OrganizationsService;
import com.mytegroup.api.service.users.UsersService;
import com.mytegroup.api.service.waitlist.WaitlistService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Service for authentication.
 * Handles login, registration, email verification, and password reset.
 * Mirrors NestJS auth.service.ts implementation.
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {
    
    private final UsersService usersService;
    private final OrganizationsService organizationsService;
    private final WaitlistService waitlistService;
    private final EmailService emailService;
    private final PasswordEncoder passwordEncoder;
    private final AuditLogService auditLogService;
    
    @Value("${app.auth.verification-token-ttl-hours:24}")
    private int verificationTokenTtlHours;
    
    @Value("${app.auth.reset-token-ttl-hours:1}")
    private int resetTokenTtlHours;
    
    /**
     * Authenticates a user with email and password.
     */
    @Transactional(readOnly = true)
    public User login(String email, String password) {
        User user = usersService.findByEmail(email);
        boolean isValid = user != null && passwordEncoder.matches(password, user.getPasswordHash());
        
        String orgId = user != null && user.getOrganization() != null 
            ? user.getOrganization().getId().toString() 
            : null;
        
        Map<String, Object> metadata = new HashMap<>();
        metadata.put("email", email);
        metadata.put("success", isValid);
        
        auditLogService.log(
            "auth.login",
            orgId,
            user != null ? user.getId().toString() : null,
            "User",
            user != null ? user.getId().toString() : null,
            metadata
        );
        
        if (!isValid) {
            throw new UnauthorizedException("Invalid credentials");
        }
        
        if (user.getArchivedAt() != null) {
            throw new ForbiddenException("User archived");
        }
        if (Boolean.TRUE.equals(user.getLegalHold())) {
            throw new ForbiddenException("User on legal hold");
        }
        if (!Boolean.TRUE.equals(user.getIsEmailVerified())) {
            throw new ForbiddenException("Email not verified");
        }
        
        usersService.markLastLogin(user.getId());
        
        return user;
    }
    
    /**
     * Registers a new user.
     */
    @Transactional
    public User register(String email, String password, String username, String firstName, 
                        String lastName, String organizationName, String orgId, 
                        boolean legalAccepted, String inviteToken) {
        
        // Validate legal acceptance
        if (!legalAccepted) {
            throw new BadRequestException("You must accept the Privacy Policy and Terms & Conditions to create an account");
        }
        
        // Validate password strength
        if (!PasswordValidator.isStrong(password)) {
            throw new BadRequestException(PasswordValidator.STRONG_PASSWORD_MESSAGE);
        }
        
        String normalizedEmail = DomainUtil.normalizeEmail(email);
        if (normalizedEmail == null) {
            throw new BadRequestException("Invalid email");
        }
        
        String domain = DomainUtil.normalizeDomainFromEmail(normalizedEmail);
        if (domain == null || domain.isEmpty()) {
            throw new BadRequestException("Invalid email domain");
        }
        
        String derivedUsername = username != null && !username.trim().isEmpty()
            ? username.trim()
            : (firstName != null && lastName != null 
                ? (firstName.trim() + " " + lastName.trim()).trim() 
                : normalizedEmail.split("@")[0]);
        
        // ==================== Waitlist/Invite Gate Logic ====================
        
        Map<String, Object> waitlistEntry = null;
        try {
            waitlistEntry = waitlistService.findByEmail(normalizedEmail);
        } catch (Exception e) {
            log.debug("Waitlist lookup failed: {}", e.getMessage());
        }
        
        boolean inviteGateEnabled = waitlistService.shouldEnforceInviteGate();
        
        if (inviteGateEnabled) {
            String status = waitlistEntry != null ? (String) waitlistEntry.get("status") : "missing";
            
            if (waitlistEntry == null || !List.of("invited", "activated").contains(status)) {
                waitlistService.logEvent("waitlist.register_blocked", Map.of(
                    "email", normalizedEmail,
                    "status", status
                ));
                throw new ForbiddenException(
                    "Invite required. Request early access, verify your email + phone, and we will invite your company in a wave."
                );
            }
            
            String verifyStatus = (String) waitlistEntry.get("verifyStatus");
            String phoneVerifyStatus = (String) waitlistEntry.get("phoneVerifyStatus");
            boolean fullyVerified = "verified".equals(verifyStatus) && "verified".equals(phoneVerifyStatus);
            
            if (!fullyVerified) {
                waitlistService.logEvent("waitlist.register_blocked_unverified", Map.of(
                    "email", normalizedEmail,
                    "verifyStatus", verifyStatus != null ? verifyStatus : "null",
                    "phoneVerifyStatus", phoneVerifyStatus != null ? phoneVerifyStatus : "null"
                ));
                throw new ForbiddenException("Please verify your email and phone before creating your account.");
            }
            
            // Check invite token if required
            if (waitlistService.requiresInviteToken()) {
                if (inviteToken == null || inviteToken.isEmpty()) {
                    throw new ForbiddenException("Invite link required. Please use the invite email to register.");
                }
                
                String inviteHash = TokenHashUtil.hashToken(inviteToken);
                String storedHash = (String) waitlistEntry.get("inviteTokenHash");
                Object expiresAt = waitlistEntry.get("inviteTokenExpiresAt");
                
                boolean tokenExpired = false;
                if (expiresAt != null) {
                    LocalDateTime expires = expiresAt instanceof LocalDateTime 
                        ? (LocalDateTime) expiresAt 
                        : LocalDateTime.parse(expiresAt.toString());
                    tokenExpired = expires.isBefore(LocalDateTime.now());
                }
                
                if (storedHash == null || !storedHash.equals(inviteHash) || tokenExpired) {
                    throw new ForbiddenException("Invite link invalid or expired. Please request a fresh invite.");
                }
            }
        }
        
        // ==================== Domain Gate Logic ====================
        
        boolean domainLocked = false;
        if (waitlistService.domainGateEnabled()) {
            Organization existingDomainOrg = organizationsService.findByDomain(domain);
            if (existingDomainOrg != null) {
                waitlistService.logEvent("waitlist.register_blocked_domain", Map.of(
                    "email", normalizedEmail,
                    "domain", domain,
                    "orgId", existingDomainOrg.getId().toString()
                ));
                throw new ForbiddenException("Your company already has access. Please ask your org admin to invite you.");
            }
            
            boolean claimed = waitlistService.acquireDomainClaim(domain);
            domainLocked = claimed;
            if (!claimed) {
                waitlistService.logEvent("waitlist.register_blocked_domain_race", Map.of(
                    "email", normalizedEmail,
                    "domain", domain
                ));
                throw new ForbiddenException("Your company already has access. Please ask your org admin to invite you.");
            }
        }
        
        try {
            // ==================== Create Organization if needed ====================
            
            Organization org;
            Long createdOrgId = null;
            
            if (orgId == null || orgId.trim().isEmpty()) {
                Organization newOrg = new Organization();
                newOrg.setName(organizationName != null ? organizationName : derivedUsername + "'s Organization");
                newOrg.setPrimaryDomain(domain);
                org = organizationsService.create(newOrg);
                createdOrgId = org.getId();
            } else {
                org = organizationsService.findById(Long.parseLong(orgId));
            }
            
            // ==================== Create User ====================
            
            // Generate email verification token if invite gate is not enforced
            TokenHashUtil.TokenData emailVerification = inviteGateEnabled 
                ? null 
                : TokenHashUtil.generateTokenWithHashHours(verificationTokenTtlHours);
            
            Role assignedRole = createdOrgId != null ? Role.ORG_OWNER : Role.USER;
            
            User user = new User();
            user.setUsername(derivedUsername);
            user.setFirstName(firstName != null ? firstName.trim() : null);
            user.setLastName(lastName != null ? lastName.trim() : null);
            user.setEmail(normalizedEmail);
            user.setPasswordHash(password); // Will be hashed in UsersService.create
            user.setOrganization(org);
            user.setRole(assignedRole);
            user.setRoles(new ArrayList<>(List.of(assignedRole)));
            user.setIsOrgOwner(createdOrgId != null);
            user.setIsEmailVerified(inviteGateEnabled); // Auto-verified if invite gate is on
            user.setVerificationTokenHash(emailVerification != null ? emailVerification.hash() : null);
            user.setVerificationTokenExpires(emailVerification != null ? emailVerification.expires() : null);
            
            User savedUser = usersService.create(user, true); // Enforce seat
            
            // Set org owner if new org was created
            if (createdOrgId != null) {
                organizationsService.setOwner(createdOrgId, savedUser.getId());
            }
            
            // Mark activated in waitlist
            String cohortTag = waitlistEntry != null ? (String) waitlistEntry.get("cohortTag") : null;
            waitlistService.markActivated(normalizedEmail, cohortTag);
            
            // Audit log
            auditLogService.log(
                "auth.register",
                org.getId().toString(),
                savedUser.getId().toString(),
                "User",
                savedUser.getId().toString(),
                null
            );
            
            // Send verification email if needed
            if (emailVerification != null) {
                try {
                    emailService.sendVerificationEmail(
                        normalizedEmail, 
                        emailVerification.token(), 
                        org.getId().toString(), 
                        derivedUsername
                    );
                } catch (Exception e) {
                    log.error("Failed to send verification email: {}", e.getMessage());
                    // Don't fail registration if email fails
                }
            }
            
            return savedUser;
            
        } finally {
            // Always release domain claim on exit
            if (domainLocked) {
                waitlistService.releaseDomainClaim(domain);
            }
        }
    }
    
    /**
     * Verifies email with token.
     */
    @Transactional
    public User verifyEmail(String token) {
        if (token == null || token.isEmpty()) {
            throw new BadRequestException("Token is required");
        }
        
        String hash = TokenHashUtil.hashToken(token);
        User user = usersService.findByVerificationToken(hash);
        
        if (user == null) {
            throw new BadRequestException("Invalid or expired token");
        }
        
        User updatedUser = usersService.clearVerificationToken(user.getId());
        
        String orgId = user.getOrganization() != null 
            ? user.getOrganization().getId().toString() 
            : null;
        
        auditLogService.log(
            "auth.verify_email",
            orgId,
            user.getId().toString(),
            "User",
            user.getId().toString(),
            null
        );
        
        return updatedUser;
    }
    
    /**
     * Initiates password reset.
     */
    @Transactional
    public Map<String, String> forgotPassword(String email) {
        User user = usersService.findByEmail(email);
        
        if (user == null) {
            // Don't leak existence
            return Map.of("status", "ok");
        }
        
        if (Boolean.TRUE.equals(user.getLegalHold()) || user.getArchivedAt() != null) {
            throw new ForbiddenException("User unavailable");
        }
        
        // Generate reset token
        TokenHashUtil.TokenData tokenData = TokenHashUtil.generateTokenWithHashHours(resetTokenTtlHours);
        
        usersService.setResetToken(user.getId(), tokenData.hash(), tokenData.expires());
        
        String orgId = user.getOrganization() != null 
            ? user.getOrganization().getId().toString() 
            : null;
        
        auditLogService.log(
            "auth.password_reset_requested",
            orgId,
            user.getId().toString(),
            "User",
            user.getId().toString(),
            null
        );
        
        // Send password reset email
        try {
            emailService.sendPasswordResetEmail(
                user.getEmail(), 
                tokenData.token(), 
                orgId, 
                user.getUsername()
            );
        } catch (Exception e) {
            log.error("Failed to send password reset email: {}", e.getMessage());
            // Still return ok to not leak info
        }
        
        return Map.of("status", "ok");
    }
    
    /**
     * Resets password with token.
     */
    @Transactional
    public User resetPassword(String token, String newPassword) {
        // Validate password strength
        if (!PasswordValidator.isStrong(newPassword)) {
            throw new BadRequestException(PasswordValidator.STRONG_PASSWORD_MESSAGE);
        }
        
        if (token == null || token.isEmpty()) {
            throw new BadRequestException("Token is required");
        }
        
        String hash = TokenHashUtil.hashToken(token);
        User user = usersService.findByResetToken(hash);
        
        if (user == null) {
            throw new BadRequestException("Invalid or expired token");
        }
        
        if (Boolean.TRUE.equals(user.getLegalHold()) || user.getArchivedAt() != null) {
            throw new ForbiddenException("User unavailable");
        }
        
        User updatedUser = usersService.clearResetTokenAndSetPassword(user.getId(), newPassword);
        
        String orgId = user.getOrganization() != null 
            ? user.getOrganization().getId().toString() 
            : null;
        
        auditLogService.log(
            "auth.password_reset",
            orgId,
            user.getId().toString(),
            "User",
            user.getId().toString(),
            null
        );
        
        return updatedUser;
    }
    
    /**
     * Checks password strength and returns analysis.
     */
    public Map<String, Object> passwordStrength(String password) {
        PasswordValidator.PasswordStrengthResult result = PasswordValidator.getStrength(password);
        
        Map<String, Object> response = new HashMap<>();
        response.put("score", result.score());
        response.put("crackTimesDisplay", result.crackTimesDisplay());
        response.put("feedback", result.feedback());
        response.put("guessesLog10", result.guessesLog10());
        
        return response;
    }
    
    /**
     * Lists users for an organization.
     */
    @Transactional(readOnly = true)
    public List<User> listUsers(String orgId) {
        if (orgId == null) {
            throw new com.mytegroup.api.exception.BadRequestException("orgId is required");
        }
        return usersService.list(orgId, false);
    }
}
